students/rhamersky/Lesson_6/test_mailroom_part4.py

```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env python3
+import unittest
+from mailroom_part4 import tup_donor_names, Menu, SubMenu, Files, update_donor, donor_list_creation,send_thankyou_option, \
+    create_report_option, exit_program_option,lifetime_donations,menu_dispatch
+
+class MyTestCase(unittest.TestCase):
+    def test_creatReport(self):
+        self.assertIsNone(SubMenu.create_report())
+
+    def test_donorEmail(self):
+        str_salutation = "Dear John Banks,"
+        str_body = "\n\n" + "Thank you for generosity! " \
+                            "\n\n" + "This donation will be put to great use!" \
+                                     "\n\n" + "Your current lifetime donation is $300.00."
+        str_valediction = "\n\n" + "Sincerely," \
+                                   "\n\n" + "The Team"
+        self.assertMultiLineEqual(Files.donor_email("John Banks",{"John Banks":"$300.00"}),
+                                  str_salutation+str_body+str_valediction)
+
+    def test_updateDonor(self):
+        self.assertIsNone(update_donor("John Banks",300.00))
+
+    def test_donorListCreation(self):
+        self.assertIsNone(donor_list_creation())
+
+    def test_sendThankyouOption(self):
+        self.assertIsNone(donor_list_creation())
+
+    def test_createReportOption(self):
+        self.assertIsNone(create_report_option())
+
+    def test_exitProgram(self):
+        self.assertEqual(exit_program_option(),"exit")
+
+    def test_lifetimeDonations(self):
+        self.assertDictEqual(lifetime_donations(tup_donor_names), lifetime_donations(tup_donor_names))
+
+    def test_menu_dispatch(self):
+        self.assertDictEqual(menu_dispatch,menu_dispatch)
+
```

Comment:
Most of these tests are not actually testing anything other than the unittest assert statements and Python itself are working as expected. Since frameworks and languages are already well tested, you never need to write these kinds of tests yourself. 

The only actual test that I can see here is `test_donorEmail` where you are testing your own program's dynamic logic against an expected result/value. 

It seems as though you may need to refactor your mailroom code so that more of your functions return testable values. This will also have the positive effect of making your code more tightly focused, flexible and extensible, which will be important as we move into more advanced lessons and mailroom assignments.

Some ways to do this would be to have your functions return strings to be printed elsewhere, rather than printing as a side effect of the function. You could also use a test donor data structure for your unit tests, and write tests for your functions that modify that data structure, with assert statements that test against expected results/values.

Reply:
I will work on this and send it back in as soon as possible.